### PR TITLE
Replace special $array function with dedicated IR

### PR DESCRIPTION
**Pull request by martint**
Function calls support a limited number of arguments, so we need dedicated support from the IR to model large arrays.

It enables arbitrarily large arrays of constant values. If elements are not constant, the size is limited by the number of bytecodes in the expression containing the array constructor (similar to any other expression containing multiple terms).

Support for arbitrarily large non-constant arrays will require splitting the construction of the array into multiple steps and methods, and should be done in a future PR.

## Release notes

(x) Release notes are required, with the following suggested text:

```markdown
# General
* Add support for large constant arrays. ({issue}`21566`)
```

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/sql/planner/IrExpressionInterpreter.java [hunk 1]
```diff
@@ -763,6 +765,24 @@ public Expression visitCast(Cast node, SymbolResolver context)
...
 
+        @Override
+        protected Expression visitArray(Array node, SymbolResolver context)
+        {
+            List<Expression> elements = node.elements().stream()
+                    .map(field -> processWithExceptionHandling(field, context))
+                    .toList();
+
+            if (elements.stream().allMatch(Constant.class::isInstance)) {
+                BlockBuilder builder = node.elementType().createBlockBuilder(null, node.elements().size());
```
Comment (wendigo) [inline core/trino-main/src/main/java/io/trino/sql/planner/IrExpressionInterpreter.java:776]: For other reviewers: here's the place where the magic happens

File: core/trino-main/src/main/java/io/trino/sql/planner/IrExpressionInterpreter.java [hunk 2]
```diff
@@ -763,6 +765,24 @@ public Expression visitCast(Cast node, SymbolResolver context)
...
+                for (Expression element : elements) {
+                    writeNativeValue(node.elementType(), builder, ((Constant) element).value());
+                }
+                return new Constant(node.type(), builder.build());
+            }
+
+            return new Array(node.elementType(), elements);
```
Comment (wendigo) [inline core/trino-main/src/main/java/io/trino/sql/planner/IrExpressionInterpreter.java:783]: If there are both constant and non-constant elements, does it make sense to represent it as the sum of arrays where one is `Constant` and second one is `Array` ? Does the order matter here?

**Review by wendigo — APPROVED** [review]

File: core/trino-main/src/main/java/io/trino/sql/gen/RowExpressionCompiler.java [hunk 3]
```diff
@@ -135,6 +135,9 @@ public BytecodeNode visitSpecialForm(SpecialForm specialForm, Context context)
                 case ROW_CONSTRUCTOR:
                     generator = new RowConstructorCodeGenerator(specialForm);
                     break;
+                case ARRAY_CONSTRUCTOR:
```
Comment (wendigo) [inline core/trino-main/src/main/java/io/trino/sql/gen/RowExpressionCompiler.java:138]: nit: this switch could be modernized to enhanced one

File: core/trino-main/src/main/java/io/trino/sql/planner/IrExpressionInterpreter.java (hunk 2 above)
Comment (martint) [reply core/trino-main/src/main/java/io/trino/sql/planner/IrExpressionInterpreter.java:783]: Yes, the order matters. In theory, we could find runs of constant and non-constant segments and split it, but I don’t think it buys us much.

File: core/trino-main/src/main/java/io/trino/sql/gen/RowExpressionCompiler.java (hunk 3 above)
Comment (martint) [reply core/trino-main/src/main/java/io/trino/sql/gen/RowExpressionCompiler.java:138]: Yeah, I’ll do it separately.

File: core/trino-main/src/main/java/io/trino/sql/planner/IrExpressionInterpreter.java (hunk 2 above)
Comment (wendigo) [reply core/trino-main/src/main/java/io/trino/sql/planner/IrExpressionInterpreter.java:783]: Sure, merge it then :)

**Review by wendigo — APPROVED** [review]

File: core/trino-main/src/main/java/io/trino/sql/planner/TranslationMap.java [hunk 4]
```diff
@@ -674,20 +672,12 @@ private io.trino.sql.ir.Expression translate(DereferenceExpression expression)
...
         List<io.trino.sql.ir.Expression> values = expression.getValues().stream()
                 .map(this::translateExpression)
                 .collect(toImmutableList());
 
-        return BuiltinFunctionCallBuilder.resolve(plannerContext.getMetadata())
-                .setName(ArrayConstructor.NAME)
-                .setArguments(types, values)
-                .build();
+        Type type = analysis.getType(expression);
+        return new io.trino.sql.ir.Array(((ArrayType) type).getElementType(), values);
```
Comment (dain) [inline core/trino-main/src/main/java/io/trino/sql/planner/TranslationMap.java:680]: Can this be imported?

File: core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java [hunk 5]
```diff
@@ -1379,14 +1374,7 @@ else if (isAggregation) {
                 throw new TrinoException(e::getErrorCode, extractLocation(node), e.getMessage(), e);
             }
 
-            if (function.getSignature().getName().equals(ARRAY_CONSTRUCTOR_NAME)) {
```
Comment (dain) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java:1382]: We might want some general catch.  If the list is not constant, I expect there is a limit between 5-10 k given the number of bytecodes to push data on the stack and to call the methods.  If it is constant, I expect there is a point where it it just too much data, althought they likely hit the 1m query limit first.

**Review by dain — APPROVED** [review]

File: core/trino-main/src/main/java/io/trino/sql/planner/TranslationMap.java (hunk 4 above)
Comment (martint) [reply core/trino-main/src/main/java/io/trino/sql/planner/TranslationMap.java:680]: No, it conflicts with Array from the AST

File: core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java (hunk 5 above)
Comment (martint) [reply core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java:1382]: It’s caught by the expression compiler, just like with any other large expression.

**Pull request merged**